Sts1CobcSw/Periphery/HardwareCrc32.cpp

```diff
@@ -0,0 +1,151 @@
+namespace sts1cobcsw::periphery
+{
+
+constexpr auto bytesPerWord = 4U;
+auto crcDma = DMA2_Stream1;
+
+
+extern "C"
+{
+void DMA1_Stream1_IRQHandler();
+}
+
+
+auto EnableHardwareCrc() -> void
+{
+    RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_CRC, ENABLE);
+}
+
+
+auto EnableHardwareCrcAndDma() -> void
+{
+    // Proposed: DMA 2 Stream 1
+
+    // Enable DMA 1
+    RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_DMA2, ENABLE);
+
+    // Initialize
+    // https://www.st.com/resource/en/application_note/an4187-using-the-crc-peripheral-on-stm32-microcontrollers-stmicroelectronics.pdf
+    DMA_InitTypeDef dmaInitStruct;
+    DMA_StructInit(&dmaInitStruct);
+
+    // Check if this is needed, seen in an online example but not in the STM manual
+    // DMA_DeInit(DMA2_Stream1);
+    dmaInitStruct.DMA_DIR = DMA_DIR_MemoryToMemory;
+
+    // Changed when there is an actual transfer
+    // dmaInitStruct.DMA_PeripheralBaseAddr = static_cast<uint32_t>(0U);
+
+    dmaInitStruct.DMA_Memory0BaseAddr = reinterpret_cast<uintptr_t>(&(CRC->DR));
+    dmaInitStruct.DMA_PeripheralInc = DMA_PeripheralInc_Enable;
+    dmaInitStruct.DMA_MemoryInc = DMA_MemoryInc_Disable;
+
+    // Changed when there is an actual transfer
+    // dmaInitStruct.DMA_BufferSize = static_cast<uint32_t>(1U);
+
+    dmaInitStruct.DMA_PeripheralDataSize = DMA_PeripheralDataSize_Word;
+
+    // TODO: Test this with byte and memory 4-burst instead of single and word
+    dmaInitStruct.DMA_MemoryDataSize = DMA_MemoryDataSize_Word;
+    dmaInitStruct.DMA_Mode = DMA_Mode_Normal;
+
+    // TODO: Test this with byte and memory 4-burst instead of single and word
+    dmaInitStruct.DMA_MemoryBurst = DMA_MemoryBurst_Single;
+    dmaInitStruct.DMA_PeripheralBurst = DMA_PeripheralBurst_Single;
+    dmaInitStruct.DMA_FIFOMode = DMA_FIFOMode_Disable;
+
+    // TODO: Check necessary priority, default is low
+    // DMA_InitStruct.DMA_Priority = DMA_Priority_something;
+
+    DMA_Init(crcDma, &dmaInitStruct);
+    DMA_ITConfig(crcDma, DMA_IT_TC, ENABLE);
+
+    // Enable CRC Peripheral
+    // TODO: Check if we do it once in the beginning, or with every CRC calculation
+    RCC_AHB1PeriphClockCmd(RCC_AHB1Periph_CRC, ENABLE);
+
+    // Enable DMA 2 Stream 1 Interrupt
+    NVIC_EnableIRQ(DMA2_Stream1_IRQn);
+}
+
+auto DmaCrc32(std::span<uint8_t> data) -> void
+{
+    // Format data byte buffer to words
+    // If not divisible by 4 bytes, 1 word padding is needed
+    auto trailingBytes = data.size() % bytesPerWord;
+    auto padding = trailingBytes == 0 ? 0U : 1U;
+    auto tmpBufferSize = data.size() / bytesPerWord + padding;
+    uint32_t tmpBuffer[tmpBufferSize];
+
+    // Careful: don't forget that trailing bytes can't be added like this
+    for(size_t i = 0; i < tmpBufferSize - padding; i++)
+    {
+        uint32_t word = 0U;
+        word |= (data[i * 4U] << CHAR_BIT * 3) | (data[i * 4U + 1] << CHAR_BIT * 2)
+              | (data[i * 4U + 2] << CHAR_BIT) | (data[i * 4U + 3]);
+        tmpBuffer[i] = word;
+    }
```

Comment:
This looks a lot like `std::memcpy`. If `data` is large, though, we cannot afford to convert bytes to words by copying. Instead, we should use something like `reinterpret_cast<std::uint32_t *>(data.data())`, but then we have to take care of possible alignment issues because not every byte array has 4-byte alignment.

I think, we need a meeting for this and the padding below.